### PR TITLE
gitlab_user: add expires_at option

### PR DESCRIPTION
**Pull request by aminvakil**
##### SUMMARY
Fixes: https://github.com/ansible-collections/community.general/issues/2325

##### ISSUE TYPE
- Feature Pull Request

##### COMPONENT NAME
plugins/modules/source_control/gitlab/gitlab_user.py

### CODE REVIEW

**Comment by felixfontein** [comment]
You also need to change `addSshKeyToUser`, otherwise this has no effect.

(I'm not sure whether the module is doing any idempotence checks for SSH keys. If it does, this needs to be added there as well.)

**Comment by aminvakil** [comment]
> You also need to change `addSshKeyToUser`, otherwise this has no effect.

Fixed in https://github.com/ansible-collections/community.general/pull/2450/commits/0b8814757ae240a6d9c2cfb28074410b086627a7.

> (I'm not sure whether the module is doing any idempotence checks for SSH keys. If it does, this needs to be added there as well.)

I think it's only relying on `sshkey_name` and if it a sshkey with name has already been created, it will not try to add ssh key, which maybe a problem for an already created ssh key which user now wants to add `expire_at` to it.

Either way I'm going to check this new option using added integration test on an instance and return with results.

File: plugins/modules/source_control/gitlab/gitlab_user.py [hunk 1]
```diff
@@ -57,16 +57,22 @@
...
+      - The name of the sshkey.
     type: str
   sshkey_file:
     description:
       - The ssh key itself.
     type: str
+  sshkey_expires_at:
+    description:
+      - The expiration date of the SSH key in ISO 8601 format C(YYYY-MM-DDTHH:MM:SSZ).
+      - It does not get added on an already present ssh key.
```
Comment (felixfontein) [inline plugins/modules/source_control/gitlab/gitlab_user.py:69]:
```suggestion
      - This is only used when adding new SSH keys.
```

File: plugins/modules/source_control/gitlab/gitlab_user.py [hunk 2]
```diff
@@ -57,16 +57,22 @@
     type: str
   sshkey_name:
     description:
-      - The name of the sshkey
+      - The name of the sshkey.
```
Comment (felixfontein) [inline plugins/modules/source_control/gitlab/gitlab_user.py:60]:
```suggestion
      - The name of the SSH public key.
```

File: plugins/modules/source_control/gitlab/gitlab_user.py [hunk 3]
```diff
@@ -305,7 +312,8 @@ def addSshKeyToUser(self, user, sshkey):
             try:
                 user.keys.create({
                     'title': sshkey['name'],
-                    'key': sshkey['file']})
+                    'key': sshkey['file'],
+                    'expires_at': sshkey['expires_at']})
```
Comment (felixfontein) [inline plugins/modules/source_control/gitlab/gitlab_user.py:316]: Is this option present with all API versions we support, or is it better to only pass it when not `None`?

Comment (aminvakil) [reply plugins/modules/source_control/gitlab/gitlab_user.py:316]: It's better to only pass it when it's not `None` for sure, also from what I can see in https://docs.gitlab.com/archives/ the oldest doc available is `10.3` version of gitlab which is only 18 months old, so I guess we don't have an option here.

Comment (aminvakil) [reply plugins/modules/source_control/gitlab/gitlab_user.py]: Fixed in https://github.com/ansible-collections/community.general/pull/2450/commits/77e3fba79d0fb535cf2444b2a6c896a5d3ebddd1.

File: plugins/modules/source_control/gitlab/gitlab_user.py [hunk 4]
```diff
@@ -295,17 +302,23 @@ def sshKeyExists(self, user, sshkey_name):
...
-                    'key': sshkey['file']})
+                if sshkey['expires_at'] is None:
+                    user.keys.create({
+                        'title': sshkey['name'],
+                        'key': sshkey['file']})
+                else:
+                    user.keys.create({
+                        'title': sshkey['name'],
+                        'key': sshkey['file'],
+                        'expires_at': sshkey['expires_at']})
```
Comment (felixfontein) [inline plugins/modules/source_control/gitlab/gitlab_user.py:321]: How about changing this to:
```suggestion
                parameter = {
                    'title': sshkey['name'],
                    'key': sshkey['file'],
                }
                if sshkey['expires_at'] is not None:
                    parameter['expires_at'] = sshkey['expires_at']
                user.keys.create(parameter)
```

**Review by felixfontein — APPROVED** [review]
I'll merge soon if nobody complains.

**Comment by felixfontein** [comment]
@aminvakil thanks for improving this module!

**Pull request merged**